### ISSUE TITLE
TODO & Wishlist

### ISSUE THREAD

**Issue by abusalimov** [abusalimov/mrcalc#1]
Here's what we have so far and what would be nice to have eventually.
## Language features
- [x] Simple expressions:
  - [x] number literals: `42`
  - [x] parens: `(((0)))`
  - [x] arithmetic operations: `(1+5)*(8+1)`
- [x] Variables: 
  - [x] definitions: `var name = 1`
  - [x] references: `foo`
- [x] Sequences: `{1, 3}`
- [x] `map`/`reduce`
  - [x] Lambdas
- [x] Output statements: `out "ret ="; print answer`
## Core structure
- [x] AST
  - [x] Node types
    - [x] expr
    - [x] stmt
  - [x] Visitor
- [x] Diagnostic (problem markers)
  - [x] Location tracking
  - [x] Listener interface for parser/compiler/interpreter
  - [x] ~~Diagnostic kind: `ERROR`/`WARNING` (_is it needed?_)~~
  - [x] ~~`secondaryLocation`? (like _note: previously defined here..._)~~
- [x] Parser interface
  - [x] ANTLR-based implementation
  - [ ] _(nice to have)_ a simple recursive descent LL(1) parser implementation with no backtracking
  - [x] Token abstraction (required for draft syntax highlighting in UI)
  - [x] `List<Token> tokenize(...)` (ditto)
- [x] Compiler
  - [x] Linkage: bind variable references to definitions
  - [x] Type inference and checks:
    - [x] integer/float distinction
    - [x] Sub-expressions
    - [x] Variable references
    - [x] `map`/`reduce` lambdas
- [x] Backend
  - [x] Function call tree
  - [x] JVM Bytecode
- [x] Interpreter
  - [x] Dumb AST walker implementation
  - [x] Delegating to `Stmt.exec()`
- [x] Runtime
  - [x] `map`/`reduce` support functions
    - [x] overload for different argument types?
  - [x] Execute in parallel
    - [x] Java 8 Streams
    - [ ] _(nice to have)_ benchmark (concurrency level? treshold?)
## UI
- [x] Syntax highlighter
  - [x] PoC (`textArea.getStyledDocument().setCharacterAttributes`)
  - [x] _(research)_ RSyntaxTextArea
- [x] Error highlighter (DefaultHighlighter.DefaultHighlightPainter?)
- [x] ~~Mark occurrences (ditto)~~
## Error diagnostics
- [x] Syntax errors
  - [x] Invalid tokens: `0$qwe`
  - [x] Parse errors: `())`, `1+`, `2  3`
- [x] Compile errors:
  - [x] Duplicate variable definition: `var foo = 1; var foo = 2;`
  - [x] Undefined variable references: `undeclared`
  - [x] Type mismatch: `1 + {1,2}`, `map(123, a -> a^2)`
- [x] Runtime errors:
  - [x] ArithmeticException?
## Project dev & maint
- [x] README
- [ ] Sub-modules
  - `core`
  - `parser.impl.antlr`
  - `cli`
  - `ui`
- [x] Travis CI
- [x] Version tags and deployment
